### ISSUE TITLE
[pose-generation] Add history/budget-aware schedules and an empirical Aria motion prior

### ISSUE THREAD

**Issue by JanDuchscherer104** [JanDuchscherer104/ARIA-NBV#70]
Parent: #67

## Outcome

Make the candidate mixture a deterministic function of the **current rollout state**, and add an empirical motion family learned from logged Aria/ASE trajectories. Early states should retain broad exploration; later states should emphasize target-direction novelty, orbit, local refinement, and controlled revisits. The resolved schedule must remain fully auditable and reproducible.

This issue improves realism and diversity without yet training an end-to-end candidate policy.

## Current repository state

- `CandidateGenerationRuntimeContext` currently carries only a target descriptor and exposes the target center.
- Candidate family counts and parameter ranges are static for every root and successor state.
- The generator does not consume remaining budget, step index, selected poses, selected target-view directions, previous family survival, or local actor-visible clearance summaries.
- Motion realism is implemented as independent hard limits on step length, height, backward displacement, and yaw.
- Raw center proposals use handcrafted blends and uniform radii; they do not reproduce the joint distribution of translational and rotational increments in logged headset motion.

Relevant owners:

- `aria_nbv/aria_nbv/pose_generation/types.py`
- `aria_nbv/aria_nbv/pose_generation/candidate_mixture.py`
- `aria_nbv/aria_nbv/pose_generation/candidate_generation_rules.py`
- `aria_nbv/aria_nbv/rollouts/replay/engine.py`
- `aria_nbv/aria_nbv/data_handling/vin_store/`
- rollout manifest and candidate provenance writers/readers

## Theory

The proposal distribution should depend on the decision state:

```math
Q_t \sim p_{\phi_0}(Q\mid s_t,e,H_t,b_t),
```

where `phi_0` is initially a fixed, versioned schedule rather than learned parameters. A static mixture wastes support because the useful balance changes with:

- remaining budget;
- target distance and scale;
- already observed target directions;
- current reconstruction support;
- local clutter and reachable baseline;
- whether previous target-aware families survived pruning.

Hestia motivates this decomposition through a “hierarchical structure to manage the high-dimensional continuous action search space.” ARIA-NBV should adopt the hierarchy—state → family allocation → bounded pose parameters—without importing Hestia's coverage reward.

Realistic motion also requires a **joint** prior. Independent maxima define a box of allowed motion but do not model correlations such as large head turns accompanying short translations or longer forward steps accompanying smaller yaw.

## Proposed runtime context

Replace the target-only runtime context with an immutable node context:

```python
@dataclass(frozen=True, slots=True)
class CandidateNodeContext:
    sample_key: str
    target_id: str
    root_variant_id: str
    candidate_replica_id: int
    step_index: int
    horizon: int
    remaining_budget: int
    current_pose_world: PoseTW
    selected_poses_world: tuple[PoseTW, ...]
    selected_shell_indices: tuple[int, ...]
    selected_center_family_ids: tuple[int, ...]
    selected_gaze_family_ids: tuple[int, ...]
    target_descriptor: TargetDescriptor
    target_direction_first_moment: Tensor | None
    target_direction_second_moment: Tensor | None
    actor_support_summary: ActorSupportSummary | None
    sampling_key: CandidateSamplingKey
```

The directional memory follows the thesis contract:

```math
\mu_t = \frac{\sum_{k<t} w_k d_k}{\sum_{k<t}w_k+\epsilon},
\qquad
M_t = \frac{\sum_{k<t}w_k d_k d_k^\top}{\sum_{k<t}w_k+\epsilon}.
```

The signed first moment distinguishes opposite approach directions; the second moment captures concentration along viewing axes.

## Resolved schedule interface

Add a pure, presentation-free resolver:

```python
class CandidateSchedule(Protocol):
    def resolve(self, context: CandidateNodeContext) -> ResolvedCandidateProfile: ...

@dataclass(frozen=True, slots=True)
class ResolvedCandidateProfile:
    contract_version: str
    schedule_name: str
    components: tuple[ResolvedCandidateComponent, ...]
    total_count: int
    resolution_inputs: dict[str, float | int | str]
    profile_hash: str
```

Requirements:

- exact component counts sum to `N_q`;
- no mutation of a shared config object;
- all inputs used by the resolver are persisted or deterministically reconstructible;
- actor profiles may use actor-visible evidence only;
- an explicitly named `oracle_schedule` may use GT-derived diagnostics for an upper-bound ablation but cannot generate deployable-policy claims.

## Initial deterministic schedule

Use a simple piecewise schedule before any learned allocator:

### Early budget

Prioritize broad reachable support:

- empirical local motion;
- forward/body-continuity anchors;
- target stand-off correction;
- turn-in-place target acquisition;
- modest orbit/lateral coverage.

### Middle budget

Prioritize angular and geometric diversity:

- target orbit;
- lateral peek;
- target-direction novelty;
- reduced generic forward allocation.

### Late budget

Prioritize refinement and uncovered directions:

- target-local refinement;
- orbit bins least represented by directional memory;
- controlled revisit/backtrack;
- smaller radii and finer gaze banks.

### State-dependent adjustments

- Far/small target: increase stand-off approach and target-framing mass.
- Near/large target: increase backoff, orbit, and turn-in-place mass.
- One-sided directional history: allocate orbit mass to the opposite side.
- Low actor-visible free-space confidence: shorten translations and increase turn-in-place.
- Repeated family collapse: reallocate only within bounded policy rules and record the trigger; do not silently relax validity.

### Data extraction

From logged trajectory poses, compute gravity-aligned local increments over time offsets corresponding to one candidate action:

```math
\xi_t = (\Delta x,\Delta y,\Delta z,\Delta\psi,\Delta\theta),
```

with roll excluded from the optimized 5-DoF action. Retain source scene, time delta, speed, and root-stage metadata for diagnostics.

### Initial model

Start with the simplest auditable model that captures correlations:

1. stratified empirical resampling from quantile bins; or
2. a small Gaussian mixture over normalized increments.

Do not introduce a neural generative model until this baseline is measured.

Persist a motion-prior artifact containing:

```text
schema/version
source manifest and split
trajectory sampling interval
frame convention
gravitational alignment rule
fitted statistics/model parameters
fit-code revision
artifact hash
```

### Use in generation

Add `EMPIRICAL_ARIA_MOTION` as one center/gaze proposal family. It is an anchor distribution, not the entire candidate table. Target-aware orbit, stand-off, peek, and scan families remain necessary because logged behavior is not optimized for the selected target.

## Knobs owned by the schedule

The resolver may vary, with persisted outputs:

- component counts;
- radius/stand-off/orbit-angle ranges;
- target-anchor allocation;
- orientation-bank allocation;
- history-novelty weight;
- candidate reservoir multiplier request;
- fixed-anchor fraction;
- empirical-prior stratum.

The resolver must not vary label semantics such as renderer resolution, crop policy, point-fusion voxel size, RRI normalization, or discount.

## Acceptance criteria

- [ ] Candidate generation consumes a typed node context containing step, budget, history, target, and sampling identity.
- [ ] A pure deterministic schedule resolves an immutable profile per state.
- [ ] Every resolved profile and its resolution inputs/hash are persisted.
- [ ] Early/middle/late states produce intentionally different allocations under tests.
- [ ] Directional memory shifts orbit/peek allocation toward underrepresented target directions.
- [ ] An empirical 5-DoF Aria motion artifact is fitted from scene-split-safe logged trajectories.
- [ ] The empirical family reproduces key joint motion statistics better than independent uniform ranges.
- [ ] Actor schedules cannot read GT mesh, GT crop, oracle RRI, or all-candidate renders.
- [ ] Oracle schedules, if implemented, are explicitly named and excluded from actor-policy claims.
- [ ] Static profiles remain supported as frozen controls.

## Required tests

- deterministic schedule resolution and profile hashing;
- component-count conservation;
- missing-modality behavior with explicit masks/defaults;
- budget/horizon boundary behavior;
- directional-moment sign and opposite-side allocation;
- frame-invariant local trajectory increments;
- motion-prior train/validation/test scene separation;
- empirical distribution tests for step length, lateral/backward rate, vertical motion, yaw, and translation–yaw correlation;
- no oracle field access in actor schedule dependency graph;
- replay reconstruction of the exact resolved profile.

## Evidence before promotion

Compare static and dynamic schedules under matched roots/replicas:

- proposal regret and best-of-60 target gain;
- family survival and selected-family entropy by step;
- target-view directional coverage by step;
- trajectory-increment distance to the logged motion prior;
- root/trajectory early termination;
- valid support and renderer cost;
- gains for early/middle/late root cohorts;
- failure cases where adaptation reduces support or collapses diversity.

## External sources

- Lu et al., *Hestia: Voxel-Face-Aware Hierarchical Next-Best-View Acquisition*: https://openaccess.thecvf.com/content/WACV2026/html/Lu_Hestia_Voxel-Face-Aware_Hierarchical_Next-Best-View_Acquisition_for_Efficient_3D_Reconstruction_WACV_2026_paper.html
- Project Aria paper for the egocentric trajectory/sensor substrate: https://arxiv.org/abs/2308.13561

## Non-goals

- This issue does not train a learned schedule or policy.
- It does not treat logged human motion as optimal NBV behavior.
- It does not adapt oracle-label semantics during generation.
- It does not use repeated roots or replicas as independent scene-level evidence.

**Comment by JanDuchscherer104** [comment]
## Research enrichment — round 2: causal scheduling, temporal conditioning, and bounded sim-to-real variation

### 1. Separate actor-causal schedule inputs from post-pruning/oracle diagnostics

The proposed state-aware schedule currently allows previous family survival/support summaries. This is safe only when the same evidence would exist at decision time.

For each resolver input declare:

```text
source_role = actor_observed | actor_predicted | oracle_audit
available_before_proposal = true | false
```

In an actor profile, do not adapt family counts from GT-mesh clearance, oracle family survival, rendered visibility, or RRI. Those can tune a **future versioned schedule offline**, but cannot dynamically resolve the current actor candidate table. Keep a separately named oracle-adaptive schedule as an upper-bound ablation.

### 2. Condition empirical motion on elapsed time and action intent

Pose increments are not comparable without their time interval. Fit

```math
p(\Delta p,\Delta\psi,\Delta\theta\mid \Delta t,\text{root stage},\text{motion mode})
```

or normalize to velocities before integrating over the candidate action duration. Otherwise varying trajectory sampling intervals create artificial step-length modes.

Required metadata:

```text
source timestamps
interpolation policy
candidate action duration
translation and angular velocity
acceleration/jerk diagnostics where available
```

Use scene-held-out fit diagnostics and avoid drawing overlapping increments from one trajectory as independent evidence.

### 3. Learn a prior over feasible *joint* increments, not independent bounds

Kuffner's treatment of SE(3) sampling and distance metrics is useful because translations and rotations require an explicit coupled measure:

https://www.ri.cmu.edu/pub_files/pub4/kuffner_james_2004_2/kuffner_james_2004_2.pdf

Compare the empirical prior against the current box/radius sampler using:

```text
translation–yaw correlation
turn-in-place mass
lateral/forward/backward ratios
vertical-motion tails
velocity and angular-velocity quantiles
nearest-neighbor distance to logged increments
held-out log likelihood or energy distance
```

A small GMM is acceptable only if its components are stable and interpretable. Stratified empirical resampling is a strong lower-complexity control.

### 4. Preserve exploration outside the behavior prior

Logged human motion is observational behavior, not an NBV optimum. A pure behavior prior may omit rare side steps or backoffs that are valuable for reconstruction.

Resolve each shell as an explicit mixture:

```text
empirical-motion anchors
semantic target-aware challengers
broad support/exploration controls
```

Report each source separately. Never infer that low-frequency logged motion is invalid.

### 5. Add bounded domain variation as a named robustness cohort

Tobin et al. motivate domain randomization by varying simulator parameters to improve transfer:

- *Domain Randomization for Transferring Deep Neural Networks from Simulation to the Real World*: https://arxiv.org/abs/1703.06907

For ARIA-NBV, use small, source-plausible perturbations around the empirical prior rather than arbitrary action randomization:

```text
pose/trajectory noise
head-height and step-scale variation
calibration/target-descriptor uncertainty
motion-speed variation
candidate timing variation
```

These are training robustness cohorts with exact perturbation provenance. Do not randomize the oracle metric, crop, or camera convention within one label corpus.

### 6. Add schedule hysteresis and minimum family floors

A state-conditioned resolver can oscillate or collapse after one poor support estimate. Add:

```text
minimum anchor counts per family
maximum count change per step
schedule hysteresis / smoothed actor support
fallback static profile
resolved trigger reasons
```

This keeps candidate support continuous across nearby states and makes failure diagnosis possible.

### 7. Directional memory needs support and uncertainty

Persist alongside the first and second moments:

```text
effective directional sample count
weight sum
age/recency
source-role counts
uncertainty or concentration
```

A zero moment with no observations is not equivalent to balanced opposite observations. Missingness must remain explicit.

### Additional acceptance criteria

- [ ] Every schedule input is classified by source role and proposal-time availability.
- [ ] Actor schedules cannot consume post-pruning GT survival or oracle gain from the current state.
- [ ] Motion prior conditions on or normalizes for elapsed time.
- [ ] Held-out scene diagnostics validate joint translation/rotation statistics.
- [ ] The shell retains explicit empirical, semantic, and broad-exploration mass.
- [ ] Domain variation is versioned, bounded, and separate from canonical oracle semantics.
- [ ] Minimum family floors, hysteresis, and static fallback prevent resolver collapse.
- [ ] Directional zero/missing/opposite-view states remain distinguishable.

**Comment by JanDuchscherer104** [comment]
## Concrete MPS calibration protocol — 0.25/0.5/1.0 s human-worn motion windows

Project Aria's official MPS trajectory contract provides the appropriate empirical source for a wearable/head-camera prior:

- open-loop trajectory: 1 kHz VIO with strong short-span local/relative accuracy, translational velocity, and angular velocity;
- closed-loop trajectory: 1 kHz gravity-aligned mapping trajectory with reduced global drift, but loop-closure corrections may reduce local accuracy over seconds.

Primary documentation:

- https://facebookresearch.github.io/projectaria_tools/gen2/technical-specs/mps/data_formats/slam/mps_trajectory
- https://facebookresearch.github.io/projectaria_tools/gen2/research-tools/dataset/pilot/tutorials/mps_loading

This suggests a specific division of labor: use **open-loop poses/dynamics for local increment statistics** and closed-loop poses for root/world alignment and long-span audit. Do not import motion magnitudes from robot-navigation benchmarks as human-worn ground truth.

### Frozen benchmark design

For each recording/trajectory, construct non-overlapping or block-resampled increments at explicit windows:

```text
Δt ∈ {0.25 s, 0.50 s, 1.00 s}
root stage ∈ {early, middle, late}
quality-score stratum
motion intent proxy ∈ {translation-dominant, turn-in-place, mixed, near-static}
```

For every increment persist:

```text
recording and trajectory identity
open/closed-loop source role
t_start, t_end, exact Δt
T_start^-1 T_end
translation in the start-device/gravity frame
yaw, pitch, roll and geodesic rotation angle
linear/angular velocity summaries
height and backward/lateral/forward components
quality scores and missing/interpolation flags
```

### Required estimates

Report per window and scene-held-out split:

- joint translation–yaw/pitch distribution;
- forward/lateral/backward ratios;
- turn-in-place mass;
- height-change and vertical-velocity tails;
- translation–rotation correlation;
- acceleration/jerk diagnostics where interpolation supports them;
- energy distance or another frozen two-sample statistic between generated candidates and held-out logged increments;
- nearest-neighbour distance from each candidate increment to the empirical support;
- fraction of candidates outside predeclared empirical quantile envelopes.

Overlapping increments from one trajectory are strongly dependent. Aggregate within recording/scene first or use block bootstrap; do not treat every 1 kHz pair as an independent sample.

### Smallest implementation

1. Add a presentation-free extractor that converts MPS trajectory rows into a versioned `WearableMotionIncrement` table.
2. Fit no learned model initially: use stratified empirical resampling and empirical quantile envelopes as controls.
3. Compare the current sampler at 0.25/0.5/1.0-second action semantics before introducing a GMM or flow.
4. Keep explicit broad/semantic exploration mass; logged human behavior is a prior, not an NBV optimum.

### Acceptance criteria

- [ ] Candidate action duration is explicit and checkpoint/manifest-bound.
- [ ] Local increments use open-loop evidence unless a validation artifact justifies another source.
- [ ] Closed-loop alignment does not silently contaminate short-span local statistics.
- [ ] Fit and evaluation recordings/scenes are disjoint.
- [ ] Current and challenger profiles are compared at all three windows with the same validity contract.
- [ ] Motion plausibility is reported separately from candidate utility and target-family support.
- [ ] No numerical robot step/turn prior is presented as empirical Aria-head motion.
